### PR TITLE
cherry-pick from dev to point-release, PR #17697

### PR DESCRIPTION
**Pull request by nick-l-o3de**
A few small improvements for the editor startup time

This is just a cherrypick of the already accepted PR 17697, from development to point-release branch.

### CODE REVIEW

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserTreeToTableProxyModel.cpp [hunk 1]
```diff
@@ -71,19 +75,50 @@ namespace AzToolsFramework
...
+            // will be to update the value if we find it, and erase the value in the other container.
+            // this will yield the minimum amount of re-hashing and memory clearing.
+            bool insertIntoTableToTree = true;
+            bool insertIntoTreeToTable = true;
+            if (auto it = m_treeToTable.find(map); it != m_treeToTable.end())
             {
-                m_tableToTree.remove(m_treeToTable.take(map));
+                int& currentValue = it.value();
+                auto elementInOtherMap = m_tableToTree.find(currentValue);
+                m_tableToTree.erase(elementInOtherMap);
```
Comment (lemonade-dm) [inline Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserTreeToTableProxyModel.cpp:98]: This should use the [key_type](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/std/hash_table.h#L678-L713) overload of erase call as the current logic doesn't check that the `elementInOtherMap` iterator exist.

The [iterator erase](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/std/hash_table.h#L674) function doesn't check if the iterator is at the end
```suggestion
                auto elementInOtherMap = m_tableToTree.erase(currentValue);
```

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserTreeToTableProxyModel.cpp [hunk 2]
```diff
@@ -71,19 +75,50 @@ namespace AzToolsFramework
...
+                currentValue = row;
+                insertIntoTreeToTable = false;
             }
-            if (m_tableToTree.contains(row))
+
+            if (auto it = m_tableToTree.find(row); it != m_tableToTree.end())
+            {
+                QPersistentModelIndex& currentValue = it.value();
+                auto elementInOtherMap = m_treeToTable.find(currentValue);
+                m_treeToTable.erase(elementInOtherMap);
```
Comment (lemonade-dm) [inline Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserTreeToTableProxyModel.cpp:107]: This should use the `key_type` overload of `hash_table::erase` or it should check that `elementInOtherMap != m_treeToTable.end()`

```suggestion
                auto elementInOtherMap = m_treeToTable.erase(currentValue);
```

**Review by lemonade-dm — APPROVED** [review]

File: Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserTreeToTableProxyModel.cpp (hunk 2 above)
Comment (nick-l-o3de) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserTreeToTableProxyModel.cpp:107]: These are not AZStd structures
```
using TreeHash = QHash<QPersistentModelIndex, int>;
        using TableMap = QMap<int, QPersistentModelIndex>;
        ``` 
they don't actually have similar operations.  erase, for example, requires an iterator, and remove does not (operates onv alue) - but erase does not re-balance the hash and remove does, etc.

Comment (lemonade-dm) [reply Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserTreeToTableProxyModel.cpp:107]: My mistake. I thought the containers was an AZStd container.

**Pull request merged**